packages/block-editor/src/components/rich-text/index.js

```diff
@@ -7,123 +7,358 @@ import { omit } from 'lodash';
+			HTML: html,
+			plainText,
+			mode,
+			tagName,
+			canUserUseUnfilteredHTML,
+		} );
+
+		if ( typeof content === 'string' ) {
+			let valueToInsert = create( { html: content } );
+
+			// If the content should be multiline, we should process text
+			// separated by a line break as separate lines.
+			if ( this.multilineTag ) {
+				valueToInsert = replace( valueToInsert, /\n+/g, LINE_SEPARATOR );
+			}
+
+			onChange( insert( value, valueToInsert ) );
+		} else if ( content.length > 0 ) {
+			if ( canReplace ) {
+				onReplace( content );
+			} else {
+				this.onSplit( value, content );
+			}
+		}
+	}
+
+	/**
+	 * Signals to the RichText owner that the block can be replaced with two
+	 * blocks as a result of splitting the block by pressing enter, or with
+	 * blocks as a result of splitting the block by pasting block content in the
+	 * instance.
+	 *
+	 * @param  {Object} record       The rich text value to split.
+	 * @param  {Array}  pastedBlocks The pasted blocks to insert, if any.
+	 */
+	onSplit( record, pastedBlocks = [] ) {
+		const {
+			onReplace,
+			onSplit,
+			__unstableOnSplitMiddle: onSplitMiddle,
+			multiline,
+		} = this.props;
+
+		if ( ! onReplace || ! onSplit ) {
+			return;
+		}
+
+		const blocks = [];
+		const [ before, after ] = split( record );
+		const hasPastedBlocks = pastedBlocks.length > 0;
+
+		// Create a block with the content before the caret if there's no pasted
+		// blocks, or if there are pasted blocks and the value is not empty.
+		// We do not want a leading empty block on paste, but we do if split
+		// with e.g. the enter key.
+		if ( ! hasPastedBlocks || ! isEmpty( before ) ) {
+			blocks.push( onSplit( toHTMLString( {
+				value: before,
+				multilineTag: multiline,
+			} ) ) );
+		}
+
+		if ( hasPastedBlocks ) {
+			blocks.push( ...pastedBlocks );
+		} else if ( onSplitMiddle ) {
+			blocks.push( onSplitMiddle() );
+		}
+
+		// If there's pasted blocks, append a block with the content after the
+		// caret. Otherwise, do append and empty block if there is no
+		// `onSplitMiddle` prop, but if there is and the content is empty, the
+		// middle block is enough to set focus in.
+		if ( hasPastedBlocks || ! onSplitMiddle || ! isEmpty( after ) ) {
+			blocks.push( onSplit( toHTMLString( {
+				value: after,
+				multilineTag: multiline,
+			} ) ) );
+		}
+
+		// If there are pasted blocks, set the selection to the last one.
+		// Otherwise, set the selection to the second block.
+		const indexToSelect = hasPastedBlocks ? blocks.length - 1 : 1;
+
+		onReplace( blocks, indexToSelect );
+	}
+
+	inputRule( value, valueToFormat ) {
+		const { onReplace } = this.props;
+
+		if ( ! onReplace ) {
+			return;
+		}
+
+		const { start, text } = value;
+		const characterBefore = text.slice( start - 1, start );
+
+		if ( ! /\s/.test( characterBefore ) ) {
+			return;
+		}
+
+		const trimmedTextBefore = text.slice( 0, start ).trim();
+		const prefixTransforms = getBlockTransforms( 'from' )
```

Comment:
It only executes when the previous character is a space. It's an improvement from `master` where it would be called for every render (and thus every character), but sure, let's see what other improvements we can make.